src/main/kotlin/org/rust/lang/core/resolve/Namespace.kt

```diff
@@ -11,10 +11,16 @@ import org.rust.lang.core.psi.ext.RsNamedElement
 import java.util.*
 
 enum class Namespace(val itemName: String) {
-    Values("value"), Types("type"), Lifetimes("lifetime"), Macros("macro")
+    Values("value"),
+    Types("type"),
+    Lifetimes("lifetime"),
+    Macros("macro"),
+    Traits("trait")
```

Comment:
Not sure that this is a good idea :) All previous namespaces are disjoints, you can have a value, a type, a lifetime and a macro in the same scope which share a name but do not conflict. It's not true for traits, because traits are in types namespace:

https://play.rust-lang.org/?gist=f8686cc9716100527bf19468e8fbbdda&version=stable

More concretely, `Namespace`'s main purpose is to direct name resolution, and, if we don't single out traits specifically during the resolution, we probably should not have this namespace variant. I suggest creating some additional mechanism for filtering items based on their types.

Reply:
What do you think about separate `Namespace` enum only for auto import with related changes? It will lead to some code duplication but I like current approach and it seems if I use current implementation of namespace and some other mechanism for traits it will be rather ugly

Reply:
Hm, I am not sure if a separate enum is required... Could one use a `(PsiElement) -> Boolean` for `expectedNamespaces`? Ie, use a lambda function as a condition, and filter by namespace (or, for trait references, by the class of `PsiElement`) inside?